corehq/apps/cleanup/management/commands/swap_duplicate_xforms.py

```diff
@@ -0,0 +1,158 @@
+                return
+
+        ids_file_path = args[0].strip()
+        log_path = args[1].strip()
+
+        num_with_1_dup = 0
+        num_with_multi_dup = 0
+        num_with_no_dup = 0
+
+        with open(log_path, "w") as log_file:
+            with open(ids_file_path, "r") as lines:
+
+                total_lines = 0
+                for _ in lines:
+                    total_lines += 1
+                lines.seek(0)
+
+                for i, (domain, bad_xform_id) in enumerate(l.split() for l in lines):
+                    self._print_progress(i, total_lines)
+                    duplicates = self.get_duplicates(domain, bad_xform_id)
+                    if len(duplicates) == 1:
+                        num_with_1_dup += 1
+                        self.swap_doc_types(log_file, bad_xform_id, list(duplicates)[0], dry_run)
+                    elif len(duplicates) > 1:
+                        num_with_multi_dup += 1
+                        self.log_too_many_dups(log_file, bad_xform_id, duplicates)
+                    else:
+                        num_with_no_dup += 1
+
+        print "Found {} forms with no duplicates".format(num_with_no_dup)
+        print "Found {} forms with one duplicate".format(num_with_1_dup)
+        print "Found {} forms with multiple duplicates".format(num_with_multi_dup)
+
+    def get_duplicates(self, domain, xform_id):
+
+        if domain not in self.dups_by_domain:
+            self.populate_dup_map(domain)
+        return self.dups_by_domain[domain].get(UUID(xform_id), {})
+
+    def populate_dup_map(self, domain):
+        self.dups_by_domain[domain] = defaultdict(set)
+        dups = XFormInstance.view(
+            'couchforms/all_submissions_by_domain',
+            startkey=[domain, 'XFormDuplicate', '2016'],
+            endkey=[domain, "XFormDuplicate", {}],
+            reduce=False,
+            include_docs=True
+        )
+        for dup in dups:
+            match = re.match('Form is a duplicate of another! \((.*)\)', dup.problem or "")
+            if match:
+                orig_id = match.groups()[0]
+                try:
+                    orig_id = UUID(orig_id)
+                except ValueError:
+                    continue
+                self.dups_by_domain[domain][orig_id].add(dup._id)
+
+    def swap_doc_types(self, log_file, bad_xform_id, duplicate_xform_id, dry_run):
+        bad_xform = XFormInstance.get(bad_xform_id)
+        duplicate_xform = XFormInstance.get(duplicate_xform_id)
+
+        # Convert the XFormInstance to an XFormDuplicate
+        bad_xform.doc_type = XFormDuplicate.__name__
+        bad_xform.problem = "Form was missing multimedia attachments. Replaced by {} on {}".format(
+            duplicate_xform_id, datetime.now().isoformat()
+        )
+        bad_xform = XFormDuplicate.wrap(bad_xform.to_json())
+
+        # Convert the XFormDuplicate to an XFormInstance
+        del duplicate_xform.problem
```

Comment:
The original spreadsheet had estimates and early numbers for how many
resubmissions we received and had anticipated. For the big domains I saw
hundreds of resubmissions on spot check, so there should be a large number .

The forms that weren't able to resubmit were quite clustered, so its
possible that they are just close in proximity in the system.

It would be good to know what the final count of replacements is (compared
to our estimates) so we can see whether we are missing anything.
On Jun 15, 2016 11:59 AM, "Noah Carnahan" <notifications@github.com> wrote:

> In corehq/apps/cleanup/management/commands/swap_duplicate_xforms.py
> <https://github.com/dimagi/commcare-hq/pull/12039#discussion_r67191155>:
>
> > +                    continue
> > +                self.dups_by_domain[domain][orig_id].add(dup._id)
> > +
> > +    def swap_doc_types(self, log_file, bad_xform_id, duplicate_xform_id, dry_run):
> > +        bad_xform = XFormInstance.get(bad_xform_id)
> > +        duplicate_xform = XFormInstance.get(duplicate_xform_id)
> > +
> > +        # Convert the XFormInstance to an XFormDuplicate
> > +        bad_xform.doc_type = XFormDuplicate.__name__
> > +        bad_xform.problem = "Form was missing multimedia attachments. Replaced by {} on {}".format(
> > +            duplicate_xform_id, datetime.now().isoformat()
> > +        )
> > +        bad_xform = XFormDuplicate.wrap(bad_xform.to_json())
> > +
> > +        # Convert the XFormDuplicate to an XFormInstance
> > +        del duplicate_xform.problem
>
> fixed here: 909cc64
> <https://github.com/dimagi/commcare-hq/pull/12039/commits/909cc6449051c82a87b11ea7eee413f9ea883f4a>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/12039/files/99d22111577f96c0bbce701adba2f73d918b2b63#r67191155>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAJdutq6-IVyPUsO8vgaEiYznN9Ylxc9ks5qMCF1gaJpZM4I0pmP>
> .
>